src/os/bluestore/BlueStore.cc

```diff
@@ -12339,22 +12341,26 @@ void BlueStore::_zoned_cleaner_thread() {
   ceph_assert(!zoned_cleaner_started);
   zoned_cleaner_started = true;
   zoned_cleaner_cond.notify_all();
-  std::deque<uint64_t> zones_to_clean;
+  auto a = dynamic_cast<ZonedAllocator*>(shared_alloc.a);
+  ceph_assert(a);
+  auto f = dynamic_cast<ZonedFreelistManager*>(fm);
+  ceph_assert(f);
   while (true) {
-    if (zoned_cleaner_queue.empty()) {
+    const auto *zones_to_clean = a->get_zones_to_clean();
+    if (!zones_to_clean) {
       if (zoned_cleaner_stop) {
 	break;
       }
       dout(20) << __func__ << " sleep" << dendl;
       zoned_cleaner_cond.wait(l);
       dout(20) << __func__ << " wake" << dendl;
     } else {
-      zones_to_clean.swap(zoned_cleaner_queue);
       l.unlock();
-      while (!zones_to_clean.empty()) {
-	_zoned_clean_zone(zones_to_clean.front());
-	zones_to_clean.pop_front();
+      for (auto zone_num : *zones_to_clean) {
+	_zoned_clean_zone(zone_num);
```

Comment:
@tchaikov This can be merged. Thanks.

@ifed01 thanks for the review. Will address the above issue in the next PR. Please review https://github.com/ceph/ceph/pull/40923 where my student tries to do a rudimentary cleaning of a zone. The purpose of the PR is just to see if he has the right idea.